src/formula.jl

```diff
@@ -54,136 +65,194 @@ function Base.show(io::IO, f::Formula)
-            ## the non-distributed arguments of the original expression.
-            for a in subex.args[2:end]
-                new_subex = copy(ex)
-                push!(new_subex.args, a)
-                ## need to recurse here, in case there are any other
-                ## distributing operations in the sub expression
-                distribute!(new_subex)
-                push!(newargs, new_subex)
-            end
-            ex.args = newargs
-        end
+
+struct Subtraction <: FormulaRewrite end
+applies(ex::Expr, child_idx::Int, ::Type{Subtraction}) =
+    is_call(ex.args[child_idx], :-)
+function rewrite!(ex::Expr, child_idx::Int, ::Type{Subtraction})
+    child = ex.args[child_idx]
+    @argcheck child.args[3] == 1 "Can only subtract 1, got $child"
+    child.args[1] = :+
+    child.args[3] = -1
+    child_idx
+end
+
+# default re-write is a no-op (go to next child)
+rewrite!(ex::Expr, child_idx::Int, ::Nothing) = child_idx+1
+
+# like `findfirst` but returns the first element where predicate is true, or
+# nothing
+function filterfirst(f::Function, a::AbstractArray)
+    idx = findfirst(f, a)
+    idx === nothing ? nothing : a[idx]
+end
+
+
+parse!(x) = parse!(x, [Subtraction, Star, AssociativeRule, Distributive])
+parse!(x, rewrites) = x
+function parse!(i::Integer, rewrites)
+    @argcheck i ∈ [-1, 0, 1] "invalid integer term $i (only -1, 0, and 1 allowed)"
+    i
+end
+function parse!(ex::Expr, rewrites::Vector)
+    @debug "parsing $ex"
+    check_call(ex)
+    # iterate over children, checking for special rules
+    child_idx = 2
+    while child_idx <= length(ex.args)
+        @debug "  ($(ex.args[1])) i=$child_idx: $(ex.args[child_idx])"
+        # depth first: parse each child first
+        parse!(ex.args[child_idx], rewrites)
+        # find first rewrite rule that applies
+        rule = filterfirst(r->applies(ex, child_idx, r), rewrites)
+        # re-write according to that rule and update the child to position rewrite next
+        child_idx = rewrite!(ex, child_idx, rule)
     end
+    @debug "done: $ex"
     ex
 end
-distribute!(a::Any) = a
 
-const associative = Set([:+,:*,:&])       # associative special operators
+# Base.copy(x::Nothing) = x
+# Base.copy(x::Symbol) = x
 
-## If the expression is a call to the function s return its arguments
-## Otherwise return the expression
-function ex_or_args(ex::Expr,s::Symbol)
-    if ex.head != :call error("Non-call expression encountered") end
-    if ex.args[1] == s
-        ## recurse in case there are more :calls of s below
-        return vcat(map(x -> ex_or_args(x, s), ex.args[2:end])...)
+function sort_terms!(ex::Expr)
+    check_call(ex)
+    if ex.args[1] ∈ ASSOCIATIVE
+        sort!(view(ex.args, 2:length(ex.args)), by=degree)
     else
-        ## not a :call to s, return condensed version of ex
-        return condense(ex)
+        # recursively sort children
+        sort_terms!.(ex.args)
```

Comment:
`foreach(sort_terms!, ex.args)` may be more efficient here since it avoids allocations from broadcasting.